### ISSUE TITLE
Feature request: Allow multiple secret backend configurations on a single Concourse at a per-team basis

### ISSUE THREAD

**Issue by itsdalmo** [concourse/concourse#1814]
# Feature Request

## Info

Deployments: 

- [Vault 0.8.3 w/dynamodb backend](https://github.com/TeliaSoneraNorge/divx-terraform-modules/tree/master/vault)
- [Concourse 3.6.0](https://github.com/TeliaSoneraNorge/divx-terraform-modules/blob/master/concourse/concourse.tf)

## What challenge are you facing?

When running Concourse for multiple teams working on different AWS accounts, we want to supply dynamic credentials to a each team using Vault. After searching I found the following comment from @vito in https://github.com/concourse/concourse/issues/1391:

>I'm pretty confused by this. The current implementation does not lock you in to using one mount path. All the path lookups are /concourse/<team>/<pipeline>/<name>. You can mount whatever you like at that path. They can even overlap; Vault allows you to mount a generic store at /concourse and then any other type of backend at /concourse/foo.

So I ran the following and got an error message:

```bash
$ vault mount -path=concourse/<team>/aws-credentials aws
* existing mount at concourse/
```

After searching a bit, I think the solution proposed above is actually not supported by Vault:

- https://github.com/hashicorp/vault/issues/2878
- https://groups.google.com/forum/#!topic/vault-tool/wENWEjLH74s

Maybe I'm misunderstanding something here, but from the looks of it then Vault is not actually supposed to support mounting anything under `concourse/`, which also means Concourse can't support anything but generic secrets (with the current implementation of Vault credentials).

So this is a feature request to actually support other backends in Concourse (primarily AWS secret backend). 

## A Modest Proposal

I'm not actually sure what to propose. I want to say multiple top level Vault mounts, but each mount does not follow the same pattern as is used for the generic backend. E.g., for the AWS Secret backend, all we can control is `<mount-name>/creds/<cred-name>` - so at best, Concourse would know to look under creds for a team name if the backend type was AWS, but we'd have no way of providing different creds for different pipelines based on path(?).

If I misunderstood, I think Concourse could use better documentation on how to actually do this. I'd be happy to submit a PR once I understand :)

#### Related question

In the [documentation](https://concourse.ci/creds.html#vault), you mention that Concourse will fix an issue with multiple values in dynamic credentials - I could not find an open issue on this here on Github, has it been fixed or is it just not tracked?

**Comment by jtarchie** [comment]
There are three different feature requests here:

* support non-generic mounts
* support multiple mounts across teams
* support AWS backend (which has already been request in #1676)

Could we narrow down the issue or separate this out into multiple issues?

**Comment by jtarchie** [comment]
What version of vault and vault-cli are you using? I was able to create multiple mount points across the path prefix.

```
○ → vault mounts
Path                     Type       Accessor            Plugin  Default TTL  Max TTL  Force No Cache  Replication Behavior  Description
concourse/               kv         kv_561adef6         n/a     system       system   false           replicated
concourse/another_team/  kv         kv_b3646b1a         n/a     system       system   false           replicated
concourse/main/          aws        aws_80cdd8c8        n/a     system       system   false           replicated
```

**Comment by itsdalmo** [comment]
Hi, sorry for the late answer:

### Vault

```
$ vault --version
Vault v0.8.3 ('6b29fb2b7f70ed538ee2b3c057335d706b6d4e36')
```

Which is the same version as the server. There is a version 0.9.0 released now, but from the [changelog](https://github.com/hashicorp/vault/blob/v0.9.0/CHANGELOG.md)  it seems they've made the rules on submounts even stricter: 

```
core: Disallow mounting underneath an existing path, not just over [GH-2919]
```

In version 0.8.3 it's possible to mount e.g. `concourse/main` and then mount `concourse/` afterwards and effectively have a submount. However, from the changelog above and links in my original post, it seems like the Vault-team does not want us to do this and that they are actively trying to stop us from doing it.

I might have misunderstood something fundamental (I'm new to Vault), but how did you manage to create submounts and what version of Vault are you on?

### Feature request

Regarding a feature request, and they way I've understood Vault/Concourse - my feature request is for **teams having access to non-generic mounts**, being able to access multiple mounts was just a suggestion off the top of my head - there might be better ways to solve this.

>support AWS backend (which has already been request in #1676)

This is incorrect, #1676 is a feature request to support a different credential provider: [AWS SSM Parameter store](http://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-paramstore.html) - it has nothing to do with Vault, and would actually be an alternative to Vault. 

In Vault, the AWS secret backend is just another "non-generic" backend.

**Comment by iuriaranda** [comment]
I wasn't sure where to put this, both this issue and #1391 are quite related. The way I see it, the current Vault integration has the following limitations:

#### We can't access Vault secrets stored in sub-paths

For example, I could have the following key/value secret backend mounted in Vault:

```
$ vault mounts
Path                               Type       Accessor            Plugin  Default TTL  Max TTL  Force No Cache  Replication Behavior  Description
concourse/my-team/a-pipeline/      kv         kv_97150b05         n/a     system       system   false           replicated            Secrets for concourse pipelines
```

Then I write something in a sub-path:

```
$ vault write concourse/my-team/a-pipeline/sub-path/something value=this-is-a-secret
```

As far as I know, with the current implementation, I can't access that value, as Concourse will look up for a key in the pipeline level (`concourse/my-team/a-pipeline/<key>`).

This is more problematic in secret backends like the AWS one, where the interesting values are always in sub-paths, like `<mount-name>/creds/<cred-name>`.

#### We can't use multiple Vault mount points in the same pipeline

As also mentioned in #1391, the current credential lookup rules in Concourse don't allow the use of multiple Vault mount points in the same pipeline, specially after Vault 0.9, which prevents the use of submounts (as stated by @itsdalmo in the above comment).

## Proposal

Allow to specify sub-paths when accessing secrets in pipelines, like `((sub-path/key))`. With this, the current lookup rules can be the same, so Concourse would look in the following paths:
- `concourse/<team-name>/<pipeline-name>/sub-path/key`
- `concourse/<team-name>/sub-path/key`

This would also allow us to use different Vault mounts in the same pipeline. For example:

```
$ vault mounts
Path                               Type       Accessor            Plugin  Default TTL  Max TTL  Force No Cache  Replication Behavior  Description
concourse/my-team/a-pipeline/aws/  aws        aws_713fff20        n/a     system       system   false           replicated
concourse/my-team/a-pipeline/kv/   kv         kv_20b19bc9         n/a     system       system   false           replicated
```

To access those mounts in a pipeline one would use `((aws/creds/something))` or `((kv/sub-path/something))`.

This proposal would resolve both the above-mentioned issues, and also makes the Vault integration much more intuitive, as it mimics how Vault is used in the API or the CLI.

**Comment by vito** [comment]
@iuriaranda I haven't used it, and it's not documented, but I'm pretty sure you can do `((foo.bar.baz.buzz))` and it'll traverse sub-paths.

**Comment by iuriaranda** [comment]
@vito I've already tried that and it doesn't work.

I have the following mount in Vault:

```
$ vault mounts
Path                               Type       Accessor            Plugin  Default TTL  Max TTL  Force No Cache  Replication Behavior  Description
concourse/skyscrapers/test-vault/  kv         kv_2ea98b5e         n/a     system       system   false           replicated
```

And the following values:
```
$ vault list concourse/skyscrapers/test-vault
Keys
----
sub/
test

$ vault list concourse/skyscrapers/test-vault/sub
Keys
----
path
```

And my pipeline definition looks like this:
```yaml
---

jobs:
  - name: test
    plan:
      - task: get-secret
        params:
          VAULT_SECRET: ((sub.path))
        config:
          image_resource:
            type: docker-image
            source:
              repository: alpine
          platform: linux
          run:
            path: env
```

With all that, the task fails (error state) with the message `Expected to find variables: sub`.
It works if I use `VAULT_SECRET: ((test))` though.

**Comment by vito** [comment]
@iuriaranda I wonder if there's ambiguity with the syntax, since `((foo.bar))` could also mean `((foo.bar.value))`. What if you explicitly use `((sub.path.value))`?

**Comment by iuriaranda** [comment]
@vito same result. It was worth trying though

**Comment by madmod** [comment]
I see three related issues with the current implementation of Vault credential management. I can migrate this to my own issue if the maintainers want, but since much of this is based on other comments on this issue I chose to put it here first to get their feedback.

1. **No form of Vault values nested more than one level deep currently works with Concourse.** Only keys directly under "/concourse/main/whatever" currently work. (Like `vault write /concourse/main/foo value=bar` and `vault write /concourse/main/is-it working=yep`.) I have verified this with each of the examples below and other tests including vault paths, nested objects, nested objects with keys named "value", and all permutations of accessing those keys from Concourse. Accessing a root path like `((foo))` or a key off a root path like `((is-it.working))` does work with my Concourse so it is configured properly.

2. Vault has multiple types of nested values while Concourse only has one, which leads to conflicts and inaceessible values.

3. Concourse ignores all other keys when a key named `"value"` is present. (Reported [here](https://github.com/concourse/concourse/issues/1517#issuecomment-324842019))

Since 1 blocks my ability to meaningfully test any of these scenarios directly with Concourse I'll focus on 2 and 3 as they are relevant to the implementation of a fix for 1.

Here are my service versions:
```bash
$ vault version
Vault v0.9.1 ('87b6919dea55da61d7cd444b2442cabb8ede8ab1')

$ fly -t default sync
version 3.8.0 already matches; skipping
```

Writing a key in Vault as a nested map:
```bash
$ vault write concourse/main/test @<(cat <<EOF
{
  "nested": {
    "some": "thing one"
  }
}
EOF
)
Success! Data written to: concourse/main/test

$ vault read concourse/main/test
Key                     Value
---                     -----
refresh_interval        768h0m0s
nested                  map[some:thing one]

```

Writing a key in Vault as a map in a sub path:
```bash
$ vault write concourse/main/test/nested @<(cat <<EOF
{
  "some": "thing two"
}
EOF
)
Success! Data written to: concourse/main/test/nested

$ vault read concourse/main/test/nested
Key                     Value
---                     -----
refresh_interval        768h0m0s
some                    thing two

```

Writing a key in Vault as a path:
```bash
$ vault write concourse/main/test/nested/some some="thing three"
Success! Data written to: concourse/main/test/nested/some


$ vault read concourse/main/test/nested/some
Key                     Value
---                     -----
refresh_interval        768h0m0s
value                   thing three

```

With the current syntax of Concourse `((test.nested.some))` it becomes impossible to access some combination of "thing one", "thing two", or "thing three".
I propose that either:
A. Vault keys, maps, and paths have their own syntax which is optional and a documented order of precedence for the current syntax. `((test/nested.some))` = "thing one", `((test/nested[some]))` = "thing two", and ((test/nested/some)) = "thing three".
B. Vault types create implcit keys which are optional and a documented order of precedence for the current syntax. `((test.map.nested.map.some))` = "thing one" and `((test.path.nested.map.some))` = "thing two", and ((test.path.nested.path.some)) = "thing three". Depending on the order of precedence chosen some of these examples would be shorter.

I personally much prefer option A because it is more intuitive being like Vault syntax and would not be ambiguous with the magic keys "map" and "path" being inaccessible.

Finally for issue 3 the scenario below would make `((test.concourse))` inaccessible. Swapping [these two return statements](https://github.com/concourse/atc/blob/01c22b4450310d141d5f5e3ec6cb498e6d99c26e/creds/vault/vault.go#L41-L51) should fix it.

```bash
$ vault write concourse/main/test @<(cat <<EOF
{
  "concourse": "cant read this because",
  "value": "just broke everything"
}
EOF
)
Success! Data written to: concourse/main/test

$ vault read concourse/main/test
Key                     Value
---                     -----
refresh_interval        768h0m0s
concourse               cant read this because
value                   just broke everything

```

I think a solution to these three issues is required in order to properly support any Vault backend. For example in order to use [AWS auth backend](https://www.vaultproject.io/docs/secrets/aws/index.html) you must be able to access `my-aws-mount/roles/myrole.secret_key` and `my-aws-mount/roles/myrole.access_key`. Also to fully support the AWS auth backend Concourse would need to support sending custom parameters like `((my-aws-mount/sts/myrole.secret_key ttl=60m))` and it would need to solve the issue mentioned in the documentation of having separate api calls for the same secrets.

Once fixes for those are implemented a partial solution to the issue of Vault not allowing nested mounts is to fake it in Concourse by creating a map with a configuration represented in a Vault generic secret at a special path. (Maybe `/concourse/config/mounts`?) Rather than re-implementing a variant of Vault's policy language Concourse could read a special Vault policy named `concourse-mounts` by default which would contain permissions limiting teams and pipelines by passing them as parameters. I'm not sure if the Vault API ignores unknown parameters though.

Alternatively Concourse could implement a Vault Auth Backend so that the identity of the worker requesting the credentials is known to Vault and a pattern like Hashicorp made for configuring per team/user policies [for GitHub](https://www.vaultproject.io/docs/auth/github.html) could be used for Concourse teams and pipelines. That way all of the RBAC could be done in Vault policies which can be audited by Sentinel.

**Comment by eamonryan** [comment]
I have multiple customers who would love this feature for CredHub to reduce duplication and maintenance of keys common to multiple Concourse teams.

**Comment by jtarchie** [comment]
@eamonryan, what does this have to do with Credhub? This is a feature request for support in vault.

**Comment by topherbullock** [comment]
I can see this becoming a generic thing we'll want to support across all secret manager implementations: supporting multiple different types of backends / configurations at a per-team level.

**Comment by itsdalmo** [comment]
Not to be a stickler, but I'm not sure the current title actually reflects what was being requested in my original post - having "multiple secret backend configurations" does not actually solve the problem of not being able to use non-generic secret backends in Vault, and to my knowledge none of the other secret backends support dynamic credentials (my use case) like Vault does. 

tldr; New title/feature request is a good, but perhaps it should be in a separate issue?

**Comment by clintjedwards** [comment]
Kicking concourse's tires for the first time, but I've used vault a bunch in other projects. I'd like to jump on the current bandwagon mentioned first here: (https://github.com/concourse/concourse/issues/1391). It seems like concourse would benefit from not only moving secret configs to a per-team basis, but also being more agnostic with how lookups are performed.

Traditionally my vault environments have secrets already in a format that is used by other applications. Reformatting them into `/prefix/{team}/blah` (or duplicating the secrets) is very close to a non starter, since many applications depend on the original format.

Instead I think it would be better for most, to have the flexibility of providing a token(or other secret method) to a team and handling the security issues of credential lookups in vault's awesome permissioning system. Enabling the use of whatever secret path people want `(secret/{anything_i_want}/{anything_you_want}/lol.username)`, while also enabling people to lockdown secrets how they like.

**Comment by clintjedwards** [comment]
Would definitely love to help out with an RFC, but not sure i'm knowledge enough about concourse. I'll definitely chip in if someone starts one.
 
Just a nit: Can't speak for other credentials managers, but I think from a vault perspective points 2 and 3 are very much related. It would be cool for concourse to allow vault tokens (or possibly app-role creds) on a per team basis and then let vault handle the permissioning on which routes can be accessed.

Vault has a great permissioning system already and this model has the added benefit of making the concourse code around this simpler(removing the path restrictions and simply just bubbling any 400 error back to the user).

**Comment by gchenuet** [comment]
Good news !
Interesting to help on the RFC definition.

Having a global path available for all users and teams should also be useful.

**Comment by vito** [comment]
I've opened https://github.com/concourse/rfcs/issues/5 so there's a more visible call for RFCs. :)

**Comment by vito** [comment]
I'm actually gonna close this now in an effort to focus on the (RF)RFC, which is what we'd actually take action on (instead of this specific issue). Thanks for the input everyone!
